### ISSUE TITLE
--incremental mode doesn't process all data

### ISSUE THREAD

**Issue by nodje** [darold/pgbadger#567]
I'm not sure I'm using it correctly, but here's what I run:

```
pgbadger --dns-resolv --maxlength 50000 -I --retention 4 --prefix '%m [%p] %q%u@%d app=%a,client=%h ' -f stderr ssh://postgres//var/log/postgresql/stmt/postgresql-stmt* -O /var/www/html/pg_reports
```
This is croned to run every hour during my work day, and once at night for now.

I can very obviously see that queries to new databases are not taking into account, and probably some other queries to existing databases.

If I run the exact same commands without `-I` then I can see the queries to the new databases that have been created.

Moreover in the process of comparing runs with and without `-I` I ran the exact same command with a `-O` to a new folder:

```
pgbadger --dns-resolv --maxlength 50000 -I --retention 4 --prefix '%m [%p] %q%u@%d app=%a,client=%h ' -f stderr ssh://postgres//var/log/postgresql/stmt/postgresql-stmt* -O /var/www/html/newreport
```
the output was made to the existing folder as: `/var/www/html/pg_reports/out.html`

**Comment by darold** [comment]
Using the incremental mode with `-I` log files entries will be compared to information stored in file `/var/www/html/pg_reports/LAST_PARSED`. If pgbadger find that the log file is before the date in this file nothing will be done. Without the `-I` this file is not taken in account and not created this is why you will always see pgbadger parsing the full file.

I have tested the same command as you and I can not reproduce the issue, you might  check the value in the LAST_PARSED file and the first date in your log. Let me know.

**Comment by nodje** [comment]
So after that post, I left it running with this command 
(crontab `0 0,7,8,9,10,11,12,13,14,15,16,17 * * *` so that I can see some updated stats during my work day)

And it keeps not processing some data for some reason.
See below how it skips 


And it keeps not processing some data for some reason.
See below how it skips most hours for the past 8 days:
![Screenshot 2020-03-05 at 17 39 16](https://user-images.githubusercontent.com/75604/76003854-f00d4500-5f08-11ea-9756-1d435ccf1704.png)

What usually triggers me into thinking that something is amiss is the graph breaks that I see on home page:
![Screenshot 2020-03-05 at 17 45 21](https://user-images.githubusercontent.com/75604/76004054-34004a00-5f09-11ea-858b-89aa852a8811.png)

**Comment by darold** [comment]
I'm not able to reproduce this issue with remote or local log files. I think there is something wrong on your pgbadger use that make it skip some log files. If you want you can add option -d to the pgbadger cron command and redirect stderr to a file: `pgbadger command 2>>/tmp/pgbadger_debug.log` during enough time to reproduce the missing log entry issue then send me the debug output.

**Comment by nodje** [comment]
crontab task doesn't produce anymore report now, I had to run it as is manually:
```
pgbadger -d -I --dns-resolv --maxlength 50000 -v --retention 4 --prefix '%m [%p] %q%u@%d app=%a,client=%h ' -f stderr ssh://postgres//var/log/postgresql/stmt/postgresql-stmt* -O /var/www/html/pg_reports/ 2>>/tmp/pgbadger.debug
```

Here's the original log 
[pgbadger.debug.xz.log](https://github.com/darold/pgbadger/files/4312030/pgbadger.debug.xz.log)
that needs to be renamed as a .xz, then you can unarchive it (but expect ~700Mb)

One very striking problem is that there's many request that are "formated" (that is on multiple lines, and this generated lots of `DEBUG: Unknown stderr line format: [...]" noise.
Would be nice btw to be able to parse these. Many come from postgres itself, but some come from applications.

Here's a version stripped from these errors:
[pgbadger.debug.log](https://github.com/darold/pgbadger/files/4312071/pgbadger.debug.log)
but it unfortunately doesn't seem to contain any "error".

Still, this doesn't modify the existing html reports at all now.
I'd need to remove the `-I` option to regenerate from scratch

**Comment by DavidPavlicek** [comment]
I'm experiencing the same behavior. It seems like bgbadger does not parse an entire log file in incremental mode. There is my command which I execute thorough cron each hour...

`/usr/bin/pgbadger -j 2 -I -X -Z $(date +%:::z) -R 24  -T "PG2" -r foo.bar  /var/log/postgresql/postgresql-10-main.log /var/log/postgresql/postgresql-10-main.log.1 --ssh-user pgbadger -O /var/www/pgbadger/pg2`

Even if logs clearly contain newer lines, pgbadger silently ignores them. For example, the log file ends at 14:19:22 but after processing that file, LAST_PARSED contains a line with 08:25:12 timestamp.

**Comment by darold** [comment]
@nodje I think that your problem is because you have set option `-d` which is normally used to filter log entries for a database. Here you don't have a database set so it takes the following `-I` as database that mean that it won't find any valuable log entry as database `-I` doesn't exists.

@DavidPavlicek I can not reproduce the issue maybe you have not the right prefix. Or looking the pgbadger path I guess that you are not using latest development code so maybe the problem have been already fixed. You might give it a try by just overriding /usr/bin/pgbadger with file from github.

**Comment by DavidPavlicek** [comment]
@darold I use this log_line_prefix: `%t [%p]: [%l-1] user=%u,db=%d,app=%a,client=%h` which is pretty standard i think...
My pgbadger version is 11.1 (latest from PGDG repo), PostgreSQL is 10.3

I just examine old reports and it looks like this behavior was introduced somewhere between pgbadger version 9.2 and 10.3. Reports start looking weird after `apt upgrade` from 9.2 to 10.3.

**Comment by nodje** [comment]
> I'm not able to reproduce this issue with remote or local log files. I think there is something wrong on your pgbadger use that make it skip some log files. If you want you can add option -d to the pgbadger cron command and redirect stderr to a file: `pgbadger command 2>>/tmp/pgbadger_debug.log` during enough time to reproduce the missing log entry issue then send me the debug output.

Wait @darold, I used `-d` as you adviced thinking it was debug mode.
Didn't realized it was something else, I actually had `-v` already.

I'm going to remove it now, but it won't solve the problem.

**Comment by darold** [comment]
@nodje right sorry,  the -v option must be use for debug. If you can post again the resulting log without the -d option if will helps to see what happen exactly.

**Comment by jwalsh2** [comment]
I am seeing similar issues where the LAST_PARSED does not seem to be working as I expected.  I am using AWS RDS, Postgres 11.  RDS does hourly logs which are rotated hourly by AWS.  The command I am using is:

`pgbadger -I -f rds bsf-prod-logs/error/postgresql.log.2020-03-11-14 -O ~/bsf-prod-logs/reports/binary/ -H ~/bsf-prod-logs/reports/html/ -v`

Output is:

DEBUG: timezone not specified, using 18000 seconds
DEBUG: Output 'html' reports will be written to out.html
DEBUG: Autodetected log format 'rds' from bsf-prod-logs/error/postgresql.log.2020-03-11-14
DEBUG: this file has already been parsed: bsf-prod-logs/error/postgresql.log.2020-03-11-14, there no new lines in this file
DEBUG: the log statistics gathering took: 0 wallclock secs ( 0.01 usr +  0.00 sys =  0.01 CPU)
DEBUG: no new entries in your log(s) since last run.
DEBUG: building reports took:  0 wallclock secs ( 0.00 usr +  0.00 sys =  0.00 CPU)
DEBUG: the total execution time took:  0 wallclock secs ( 0.01 usr +  0.00 sys =  0.01 CPU)

The LAST_PARSED is 84K in size and contains:

2020-03-11 13:34:37     1048629 2020-03-11 13:34:37 UTC:172.31.140.246(35436):core_api@bsf:[26210]:DETAIL:  parameters: $1 = <snip out the query> ...

If I remove the LAST_PARSED, I can run it fine.

Do I need to concatenate the 24 logs into a single daily log file and then run it?

I am running the latest from the `master` branch pgbadger version 11.2

**Comment by jwalsh2** [comment]
@darold with the latest commit to master and adding in `-p p %t:%r:%u@%d:[%p]:` to the command it appears that LAST_PARSED is working as expected.  I will know more in the next 12 hours when my crons run and update the logs.

**Comment by darold** [comment]
Normally prefix  `%t:%r:%u@%d:[%p]:`  is auto-detected so it don't need to be specified but in some condition (too few lines in the log) the log format could fail. In this case I would recommend to use `-f rds` instead of the prefix.

**Comment by nodje** [comment]
I still have some logs that are skipped, even though the LAST_PARSED content seem to be accurate.
`-v` doesn't seem to produce any interesting detail, just `Parsed` and `generating` log parts.

Is there any that could help in `master` compared to `v11.1` ?

**Comment by jwalsh2** [comment]
@darold I did have the `-f rds` flag set, but I needed to had the `-p p %t:%r:%u@%d:[%p]:` flag to enable LAST_PARSED to not be ignored.

I know it seems to be a redundant setting, but since RDS sometimes rotates the logs mid query, wonder if that is what LAST_PARSED is off without the implicit setting of `-p p %t:%r:%u@%d:[%p]:`

**Comment by darold** [comment]
I do not understand, your command line prefix should be `-p '%t:%r:%u@%d:[%p]:'` here I don't understand why there is and extra `p` after `-p` and I'm surprised that it works because it should only takes `p` as prefix not the following characters.

**Comment by jwalsh2** [comment]
Sorry that was a typo with the extra `p` @darold 

Working CLI is:

`pgbadger -I edw-prod-reporting-logs/error/postgres* -O ~/edw-prod-reporting-logs/reports/binary/ -H ~/edw-prod-reporting-logs/reports/html/ -f rds -j 2 -U pgwatch2 -p %t:%r:%u@%d:[%p]: --exclude-client 172.31.53.34`

**Comment by darold** [comment]
I have tested the exact same command without being able to reproduce the problem. Probably the problem comes from you cron command that do not escape the % sign:
```
Percent-signs (%) in the command,  unless  escaped  with  backslash  (\),  will  be
changed into newline characters, and all data after the first % will be sent to the
command as standard input.
```
But as pgBadger knows the prefix for RDS logs, just using `-f rds` shall be enough, as I said you can remove the prefix setting with the `-p` option it is useless.

**Comment by nodje** [comment]
I haven't been able to put more time into this, but the problem persists, I still have only a part of the logs reported.
It's not the % escape problem for Cron.
It there's something I could add for now, it's that it seems I started to have problem after I enabled logrotate on postgres logs. Before it seemed to work fine

**Comment by darold** [comment]
This issue is probably the same as #594 and could be fixed by commit 2e14858.
